plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/server/build/N4JSProjectBuilder.java

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright (c) 2020 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.ide.server.build;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.List;
+import java.util.Set;
+
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.n4js.N4JSGlobals;
+import org.eclipse.n4js.ide.xtext.server.build.IBuildRequestFactory;
+import org.eclipse.n4js.ide.xtext.server.build.ProjectBuilder;
+import org.eclipse.n4js.ide.xtext.server.build.XBuildResult;
+import org.eclipse.n4js.projectModel.IN4JSCore;
+import org.eclipse.n4js.projectModel.IN4JSProject;
+import org.eclipse.n4js.projectModel.names.N4JSProjectName;
+import org.eclipse.n4js.tester.TestCatalogSupplier;
+import org.eclipse.n4js.xtext.workspace.XIProjectConfig;
+import org.eclipse.xtext.resource.IResourceDescription.Delta;
+import org.eclipse.xtext.util.CancelIndicator;
+
+import com.google.inject.Inject;
+
+/**
+ * Adds {@link #writeTestCatalog()} and {@link #removeTestCatalog()}.
+ */
+@SuppressWarnings("restriction")
+public class N4JSProjectBuilder extends ProjectBuilder {
+	private static final Logger LOG = LogManager.getLogger(N4JSProjectBuilder.class);
+
+	@Inject
+	private IN4JSCore n4jsCore;
+
+	@Inject
+	private TestCatalogSupplier testCatalogSupplier;
+
+	@Override
+	protected XBuildResult doBuild(IBuildRequestFactory buildRequestFactory, Set<URI> dirtyFiles, Set<URI> deletedFiles,
+			List<Delta> externalDeltas, CancelIndicator cancelIndicator) {
+
+		XBuildResult buildResult = super.doBuild(buildRequestFactory, dirtyFiles, deletedFiles, externalDeltas,
+				cancelIndicator);
+
+		writeTestCatalog();
+
+		return buildResult;
+	}
+
+	@Override
+	protected void doClearWithoutNotification() {
+		super.doClearWithoutNotification();
+
+		removeTestCatalog();
```

Comment:
Removing the test catalog is similar to deleting the project state file or deleting the output files. Since none of those two things are done in `doClearWithoutNotification()` but instead in `doClean()` directly, I think it would be nicer to override `doClean()` for invoking `removeTestCatalog()`.